WalletWasabi.Tests/UnitTests/Crypto/ProofSystemTests.cs

```diff
@@ -41,6 +44,76 @@ public void CanProveAndVerifyMAC()
 			var clientStatement = ProofSystem.IssuerParameters(coordinatorParameters, mac, Ma);
 			var isValidProof = ProofSystem.Verify(clientStatement, proofOfMac);
 
+			Assert.True(isValidProof);
+			////////
+
+			var corruptedResponses = new ScalarVector(proofOfMac.Responses.Reverse());
+			var invalidProofOfMac = new Proof(proofOfMac.PublicNonces, corruptedResponses);
+			isValidProof = ProofSystem.Verify(clientStatement, invalidProofOfMac);
+
+			Assert.False(isValidProof);
+
+			var corruptedPublicNonces = new GroupElementVector(proofOfMac.PublicNonces.Reverse());
+			invalidProofOfMac = new Proof(corruptedPublicNonces, proofOfMac.Responses);
+			isValidProof = ProofSystem.Verify(clientStatement, invalidProofOfMac);
+
+			Assert.False(isValidProof);
+		}
+
+		[Fact]
+		[Trait("UnitTest", "UnitTest")]
+		public void CanProveAndVerifySerialNumbers()
+		{
+			var rnd = new SecureRandom();
+			var z = rnd.GetScalar();
+			var r = rnd.GetScalar();
+			var a = rnd.GetScalar();
+
+			var witness = new ScalarVector(z, a, r);
+			var Ca = z * Generators.Ga + r * Generators.Gh + a * Generators.Gg;
+			var S = r * Generators.Gs;
+
+			var statement = ProofSystem.SerialNumber(Ca, S);
+			var proofOfSerialNumber = ProofSystem.Prove(new Knowledge(statement, witness), rnd);
+
+			// The coordinator must verify the blinded amout is zero
+			var isValidProof = ProofSystem.Verify(statement, proofOfSerialNumber);
+
+			Assert.True(isValidProof);
+		}
+
+		[Fact]
+		[Trait("UnitTest", "UnitTest")]
+		public void CanProveAndVerifyMacShow()
+		{
+			var rnd = new SecureRandom();
+			var coordinatorKey = new CoordinatorSecretKey(rnd);
+			var coordinatorParameters = coordinatorKey.ComputeCoordinatorParameters();
+
+			// A blinded amount is known as an `attribute`. In this case the attribute Ma is the
+			// value 10000 blinded with a random `blindingFactor`. This attribute is sent to
+			// the coordinator.
+			var amount = new Scalar(10_000);
+			var blindingFactor = rnd.GetScalar();
+			var Ma = amount * Generators.G + blindingFactor * Generators.Gh;
+
+			// The coordinator generates a MAC and a proof that the MAC was generated using the
+			// coordinator's secret key. The coordinator sends the pair (MAC + proofOfMac) back
+			// to the client.
+			var t = rnd.GetScalar();
+			var mac = MAC.ComputeMAC(coordinatorKey, Ma, t);
+
+			// The client randomizes the commitments before presenting the them to the coordinator
+			// proves to the coordinator that a credential is valid
+			var z = rnd.GetScalar();
+			var (knowledge, randomizedCommitments) = ProofSystem.MacShow(coordinatorParameters, mac, z, amount, blindingFactor);
```

Comment:
i'm not sure i chose a good argument order here, any thoughts on how to improve?

also i would rename `blindingFactor` to `r` even though it's less idiomatic C#, just so that it's easier to compare to the paper, consistent with `Ma`, `t`, and `z`

Reply:
Done.